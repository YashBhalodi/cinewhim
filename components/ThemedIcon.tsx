import { Text, type TextProps, StyleSheet } from "react-native";

import { Colors } from "@/constants/Colors";
import { FontAwesome } from "@expo/vector-icons";
import { ComponentProps } from "react";

export type ThemedIconProps = Omit<
  ComponentProps<typeof FontAwesome>,
  "color" | "size"
> & {
  variant?:
    | "iconDefault"
    | "iconSubtle"
    | "iconRed"
    | "iconNeutral"
    | "iconGreen";
  size?: "xs" | "sm" | "md" | "lg" | "xlg";
};

const SIZE_MAP = {
  xs: 12,
  sm: 16,
  md: 20,
  lg: 24,
  xlg: 28,
};

export default function ThemedIcon({
  variant = "iconDefault",
  size = "md",
  ...rest
}: ThemedIconProps) {
  const color = Colors.dark[variant];

  return <FontAwesome {...rest} color={color} size={SIZE_MAP[size]} />;
}
