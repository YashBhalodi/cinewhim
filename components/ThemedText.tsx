import { Text, type TextProps, StyleSheet } from "react-native";

import { Colors } from "@/constants/Colors";

export type ThemedTextProps = TextProps & {
  type?:
    | "body1"
    | "body2"
    | "body3"
    | "heading1"
    | "heading2"
    | "heading3"
    | "label1"
    | "label2"
    | "label3";
  color?:
    | "textDefault"
    | "textSubtle"
    | "textRed"
    | "textGreen"
    | "textNeutral";
};

export function ThemedText({
  style,
  type = "body1",
  color = "textDefault",
  ...rest
}: ThemedTextProps) {
  return (
    <Text
      style={[
        { color: Colors.dark[color] },
        styles[type] ? styles[type] : undefined,
        style,
      ]}
      {...rest}
    />
  );
}

const styles = StyleSheet.create({
  body1: {
    fontSize: 16,
    lineHeight: 24,
  },
  body2: {
    fontSize: 14,
    lineHeight: 20,
  },
  body3: {
    fontSize: 12,
    lineHeight: 16,
  },
  heading1: {
    fontSize: 28,
    lineHeight: 36,
    fontWeight: 700,
  },
  heading2: {
    fontSize: 24,
    lineHeight: 32,
    fontWeight: 700,
  },
  heading3: {
    fontSize: 20,
    lineHeight: 28,
    fontWeight: 700,
  },
  label1: {
    fontSize: 16,
    lineHeight: 24,
    fontWeight: 600,
  },
  label2: {
    fontSize: 14,
    lineHeight: 20,
    fontWeight: 600,
  },
  label3: {
    fontSize: 12,
    lineHeight: 16,
    fontWeight: 600,
  },
});
