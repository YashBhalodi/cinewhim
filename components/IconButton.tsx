import { View, Text, Pressable, StyleSheet } from "react-native";
import React, { ComponentProps, useCallback } from "react";
import { ThemedIcon, ThemedIconProps } from "./ThemedIcon";
import Animated, {
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from "react-native-reanimated";

const AnimatedPressable = Animated.createAnimatedComponent(Pressable);

interface IconButtonProps {
  iconProps: ThemedIconProps;
  onPress?: () => void;
}

const IconButton = (props: IconButtonProps) => {
  const { onPress, iconProps } = props;
  const scale = useSharedValue<number>(1);
  const animatedStyle = useAnimatedStyle(
    () => ({
      transform: [{ scale: withTiming(scale.value) }],
    }),
    []
  );

  const onPressIn = useCallback(() => {
    scale.value = 0.9;
  }, []);
  const onPressOut = useCallback(() => {
    scale.value = 1.0;
  }, []);

  return (
    <AnimatedPressable
      onPress={onPress}
      onPressIn={onPressIn}
      onPressOut={onPressOut}
      style={[styles.container, animatedStyle]}
    >
      <ThemedIcon {...iconProps} />
    </AnimatedPressable>
  );
};

const styles = StyleSheet.create({
  container: {
    borderColor: "red",
    alignItems: "center",
    justifyContent: "center",
    height: 36,
    aspectRatio: 1,
    borderRadius: 18,
    backgroundColor: "rgba(0,0,0,0.3)",
  },
});

export default IconButton;
