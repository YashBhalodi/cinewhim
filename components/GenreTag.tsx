import { StyleSheet } from "react-native";
import React from "react";
import { Genre } from "@/api/model";
import ThemedText from "./ThemedText";
import ThemedView from "./ThemedView";

interface GenreTagProps {
  genre: Genre;
}

const GenreTag = (props: GenreTagProps) => {
  const { genre } = props;

  return (
    <ThemedView style={styles.container}>
      <ThemedText>{genre.name}</ThemedText>
    </ThemedView>
  );
};

const styles = StyleSheet.create({
  container: {
    paddingHorizontal: 12,
    paddingVertical: 4,
    borderRadius: 8,
  },
});

export default GenreTag;
