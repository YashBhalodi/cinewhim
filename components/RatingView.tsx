import { View, Text, StyleSheet } from "react-native";
import React from "react";
import { ThemedIcon } from "./ThemedIcon";
import { ThemedView } from "./ThemedView";
import { ThemedText } from "./ThemedText";

interface RatingViewProps {
  rating: number;
}

const RATING_STYLE_MAP = {
  good: {
    iconColor: "iconGreen" as "iconGreen",
    textColor: "textGreen" as "textGreen",
  },
  neutral: {
    iconColor: "iconNeutral" as "iconNeutral",
    textColor: "textNeutral" as "textNeutral",
  },
  bad: {
    iconColor: "iconRed" as "iconRed",
    textColor: "textRed" as "textRed",
  },
};

const RatingView = (props: RatingViewProps) => {
  const { rating } = props;

  const ratingType = rating > 7 ? "good" : rating < 5 ? "bad" : "neutral";

  return (
    <ThemedView style={styles.container}>
      <ThemedIcon
        name="star"
        size="xs"
        variant={RATING_STYLE_MAP[ratingType].iconColor}
      />
      <ThemedText type="body3" color={RATING_STYLE_MAP[ratingType].textColor}>
        {rating.toPrecision(2)}
      </ThemedText>
    </ThemedView>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    alignItems: "center",
    gap: 4,
    borderRadius: 4,
    paddingHorizontal: 8,
    paddingVertical: 4,
    borderWidth: 1,
    backgroundColor: "rgba(0,0,0,0.6)",
    borderColor: "rgba(256,256,256,0.3)",
  },
});

export default RatingView;
