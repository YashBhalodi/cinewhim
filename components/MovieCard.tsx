import { getImageUrl } from "@/api/api";
import { Movie } from "@/api/model";
import { ImageBackground } from "expo-image";
import React, { useState } from "react";
import { Pressable, StyleSheet, View } from "react-native";
import Animated, {
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from "react-native-reanimated";
import RatingView from "./RatingView";
import { ThemedText } from "./ThemedText";
import { getReleaseDateText } from "@/utils/formattingUtils";

interface MovieCardProps {
  movie: Movie;
  onPress?: () => void;
}

const MovieCard = (props: MovieCardProps) => {
  const { movie, onPress } = props;

  const scale = useSharedValue<number>(1);
  const opacity = useSharedValue<number>(1);

  const containerAnimatedStyle = useAnimatedStyle(
    () => ({
      borderRadius: 8,
      overflow: "hidden" as "hidden",
      transform: [{ scale: withTiming(scale.value, { duration: 100 }) }],
    }),
    []
  );

  const titleContainerAnimatedStyle = useAnimatedStyle(
    () => ({
      opacity: withTiming(opacity.value, { duration: 300 }),
    }),
    []
  );

  const onPressIn = () => {
    scale.value = 0.95;
  };
  const onPressOut = () => {
    scale.value = 1;
  };

  return (
    <Pressable
      style={styles.outerContainer}
      onPressIn={onPressIn}
      onPressOut={onPressOut}
      onPress={onPress}
    >
      <Animated.View style={containerAnimatedStyle}>
        <ImageBackground
          style={styles.posterImage}
          source={getImageUrl(movie.poster_path)}
          contentFit="cover"
          contentPosition="left center"
          onLoadEnd={() => {
            opacity.value = 0;
          }}
          onError={() => {
            opacity.value = 1;
          }}
        >
          <View style={styles.ratingContainer}>
            <RatingView rating={movie.vote_average} size="sm" />
          </View>
          <View style={styles.dateContainer}>
            <ThemedText type="label3" color={"textSubtle"}>
              {getReleaseDateText(movie.release_date)}
            </ThemedText>
          </View>

          <Animated.View
            style={[styles.titleContainer, titleContainerAnimatedStyle]}
          >
            <ThemedText type="heading3">{movie.title}</ThemedText>
          </Animated.View>
        </ImageBackground>
      </Animated.View>
    </Pressable>
  );
};

const styles = StyleSheet.create({
  outerContainer: { borderRadius: 8, overflow: "hidden" },
  posterImage: {
    aspectRatio: 1 / 1.5,
    flexDirection: "column",
    justifyContent: "flex-end",
  },
  ratingContainer: {
    position: "absolute",
    top: 8,
    right: 8,
  },
  dateContainer: {
    position: "absolute",
    top: 8,
    left: 8,
    borderRadius: 4,
    padding: 4,
    borderWidth: 1,
    backgroundColor: "rgba(0,0,0,0.6)",
    borderColor: "rgba(256,256,256,0.3)",
  },
  titleContainer: {
    paddingHorizontal: 8,
    paddingVertical: 8,
    backgroundColor: "rgba(0,0,0,0.6)",
    gap: 4,
  },
});

export default MovieCard;
