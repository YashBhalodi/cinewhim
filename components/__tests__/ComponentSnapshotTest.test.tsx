import * as React from "react";
import renderer from "react-test-renderer";

import ThemedIcon from "../ThemedIcon";
import ThemedText from "../ThemedText";
import GenreTag from "../GenreTag";
import IconButton from "../IconButton";
import MovieCard from "../MovieCard";
import RatingView from "../RatingView";

describe("ThemedIcon", () => {
  it(`renders correctly`, () => {
    const tree = renderer.create(<ThemedIcon name="home" />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe("ThemedText", () => {
  it(`renders correctly`, () => {
    const tree = renderer
      .create(<ThemedText>Snapshot test!</ThemedText>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe("GenreTag", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(<GenreTag genre={{ id: 1, name: "Family" }} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("IconButton", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(<IconButton iconProps={{ name: "home" }} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("MovieCard", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <MovieCard
          movie={{
            adult: false,
            backdrop_path: "/buawWBeKYjYfeiPoS2jIcjOrghZ.jpg",
            genre_ids: [27, 53],
            id: 1214509,
            original_language: "en",
            original_title: "In a Violent Nature",
            overview:
              "The enigmatic resurrection, rampage, and retribution of an undead monster in a remote wilderness unleashes an iconic new killer after a locket is removed from a collapsed fire tower that entombed its rotting corpse.",
            popularity: 463.811,
            poster_path: "/hPfWHgq07nXbeldwEGxWB4JqwtE.jpg",
            release_date: "2024-05-31",
            title: "In a Violent Nature",
            video: false,
            vote_average: 5.794,
            vote_count: 126,
          }}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("RatingView", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(<RatingView rating={5.794} voteCount={126} size="md" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
