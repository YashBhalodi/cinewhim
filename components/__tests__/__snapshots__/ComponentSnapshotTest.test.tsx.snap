// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenreTag renders correctly 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#151718",
      },
      {
        "borderRadius": 8,
        "paddingHorizontal": 12,
        "paddingVertical": 4,
      },
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "#ECEDEE",
        },
        {
          "fontSize": 16,
          "lineHeight": 24,
        },
        undefined,
      ]
    }
  >
    Family
  </Text>
</View>
`;

exports[`IconButton renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  jestAnimatedStyle={
    {
      "value": {
        "transform": [
          {
            "scale": 1,
          },
        ],
      },
    }
  }
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "aspectRatio": 1,
      "backgroundColor": "rgba(0,0,0,0.3)",
      "borderColor": "red",
      "borderRadius": 18,
      "height": 36,
      "justifyContent": "center",
      "transform": [
        {
          "scale": 1,
        },
      ],
    }
  }
>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      [
        {
          "color": "white",
          "fontSize": 20,
        },
        undefined,
        {
          "fontFamily": "FontAwesome",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        {},
      ]
    }
  >
    
  </Text>
</View>
`;

exports[`MovieCard renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "borderRadius": 8,
      "overflow": "hidden",
    }
  }
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {
          "borderRadius": 8,
          "overflow": "hidden",
          "transform": [
            {
              "scale": 1,
            },
          ],
        },
      }
    }
    style={
      {
        "borderRadius": 8,
        "overflow": "hidden",
        "transform": [
          {
            "scale": 1,
          },
        ],
      }
    }
  >
    <View
      style={
        {
          "aspectRatio": 0.6666666666666666,
          "flexDirection": "column",
          "justifyContent": "flex-end",
        }
      }
    >
      <ViewManagerAdapter_ExpoImage
        bottom={0}
        containerViewRef={"[React.ref]"}
        contentFit="cover"
        contentPosition={
          {
            "left": 0,
            "top": "50%",
          }
        }
        left={0}
        nativeViewRef={"[React.ref]"}
        onError={[Function]}
        onLoad={[Function]}
        onLoadEnd={[Function]}
        onLoadStart={[Function]}
        onProgress={[Function]}
        placeholder={[]}
        position="absolute"
        right={0}
        source={
          [
            {
              "uri": "https://image.tmdb.org/t/p/original//hPfWHgq07nXbeldwEGxWB4JqwtE.jpg",
            },
          ]
        }
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        top={0}
        transition={null}
      />
      <View
        style={
          {
            "position": "absolute",
            "right": 8,
            "top": 8,
          }
        }
      >
        <View
          style={
            [
              {
                "backgroundColor": "#151718",
              },
              {
                "alignItems": "center",
                "backgroundColor": "rgba(0,0,0,0.6)",
                "borderColor": "rgba(256,256,256,0.3)",
                "borderRadius": 4,
                "borderWidth": 1,
                "flexDirection": "row",
                "gap": 4,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
            ]
          }
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": "#d0d4cf",
                  "fontSize": 12,
                },
                undefined,
                {
                  "fontFamily": "FontAwesome",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            
          </Text>
          <Text
            style={
              [
                {
                  "color": "#d0d4cf",
                },
                {
                  "fontSize": 12,
                  "lineHeight": 16,
                },
                undefined,
              ]
            }
          >
            5.8
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "backgroundColor": "rgba(0,0,0,0.6)",
            "borderColor": "rgba(256,256,256,0.3)",
            "borderRadius": 4,
            "borderWidth": 1,
            "left": 8,
            "padding": 4,
            "position": "absolute",
            "top": 8,
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#d4d7d9",
              },
              {
                "fontSize": 12,
                "fontWeight": 600,
                "lineHeight": 16,
              },
              undefined,
            ]
          }
        >
          May 31, 2024
        </Text>
      </View>
      <View
        collapsable={false}
        jestAnimatedStyle={
          {
            "value": {
              "opacity": 1,
            },
          }
        }
        style={
          {
            "backgroundColor": "rgba(0,0,0,0.6)",
            "gap": 4,
            "opacity": 1,
            "paddingHorizontal": 8,
            "paddingVertical": 8,
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#ECEDEE",
              },
              {
                "fontSize": 20,
                "fontWeight": 700,
                "lineHeight": 28,
              },
              undefined,
            ]
          }
        >
          In a Violent Nature
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`RatingView renders correctly 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#151718",
      },
      {
        "alignItems": "center",
        "backgroundColor": "rgba(0,0,0,0.6)",
        "borderColor": "rgba(256,256,256,0.3)",
        "borderRadius": 4,
        "borderWidth": 1,
        "flexDirection": "row",
        "gap": 4,
        "paddingHorizontal": 8,
        "paddingVertical": 4,
      },
    ]
  }
>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      [
        {
          "color": "#d0d4cf",
          "fontSize": 16,
        },
        undefined,
        {
          "fontFamily": "FontAwesome",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        {},
      ]
    }
  >
    
  </Text>
  <Text
    style={
      [
        {
          "color": "#d0d4cf",
        },
        {
          "fontSize": 14,
          "lineHeight": 20,
        },
        undefined,
      ]
    }
  >
    5.8
  </Text>
</View>
`;

exports[`ThemedIcon renders correctly 1`] = `<Text />`;

exports[`ThemedText renders correctly 1`] = `
<Text
  style={
    [
      {
        "color": "#ECEDEE",
      },
      {
        "fontSize": 16,
        "lineHeight": 24,
      },
      undefined,
    ]
  }
>
  Snapshot test!
</Text>
`;
